go/types/set_iterator.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+package types
+
+import "github.com/attic-labs/noms/go/d"
+
+type SetIterator interface {
+	// Next returns nil, when no objects remain
+	Next() Value
+
+	// AdvanceTo(v) returns:
+	// * nil if there are no objects >= v.
+	// * the next value in the iteration if  v <= curValue
+	// * the first value > v if v is not in the set, or nil if there is no value > v
+	AdvanceTo(v Value) Value
+}
+
+type setIterator struct {
+	s      Set
+	cursor *sequenceCursor
+}
+
+func (s Set) Iterator() SetIterator {
+	return &setIterator{s: s, cursor: nil}
+}
+
+func (si *setIterator) Next() (v Value) {
+	if si.cursor == nil {
+		si.cursor = newCursorAt(si.s.seq, emptyKey, false, false)
+	} else {
+		si.cursor.advance()
+	}
+	if si.cursor.valid() {
+		return si.cursor.current().(Value)
+	}
+	return nil
+}
+
+func (si *setIterator) AdvanceTo(v Value) (res Value) {
+	d.Chk.NotNil(v, "setIterator.AdvanceTo() called with nil value")
+	first := false
+	if si.cursor == nil {
+		first = true
+		si.cursor, _ = si.s.getCursorAtValue(v)
+	}
+
+	if !si.cursor.valid() {
+		return nil
+	}
+
+	curValue := si.cursor.current().(Value)
+	if first {
+		return curValue
+	}
+
+	if v.Less(curValue) {
+		return si.Next()
+	}
+
+	si.cursor, _ = si.s.getCursorAtValue(v)
+	if si.cursor.valid() {
+		return si.cursor.current().(Value)
+	}
+	return nil
+}
+
+// iterState contains iterator and it's current value
+type iterState struct {
+	i SetIterator
+	v Value
+}
+
+func (st *iterState) Next() Value {
+	if st.v == nil {
+		return nil
+	}
+	v := st.v
+	st.v = st.i.Next()
+	return v
+}
+
+func (st *iterState) AdvanceTo(v Value) Value {
+	if st.v == nil || v == nil {
+		st.v = nil
+		return nil
+	}
+	st.v = st.i.AdvanceTo(v)
+	return st.v
+}
+
+type UnionIterator struct {
+	aState iterState
+	bState iterState
+}
+
+func NewUnionIterator(iterA, iterB SetIterator) SetIterator {
+	if iterA == nil || iterB == nil {
+		if iterA == nil && iterB == nil {
+			return nil
+		}
+		if iterA == nil {
+			return iterB
+		}
+		return iterA
+	}
+	a := iterState{i: iterA, v: iterA.Next()}
```

Comment:
Whatever gets supported on SetIterator would need to be supported by Union and Intersection as well.